### ISSUE TITLE
tf random gives the same sequence of numbers when it shouldn't

### ISSUE THREAD

**Issue by laplacericky** [tensorflow/tensorflow#49917]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.5 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip3 install
- TensorFlow version (use command below): 2.5.0
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA11.3 cudnn 8.2.0
- GPU model and memory: NVIDIA TITAN V 12 GB

**Describe the current behavior**
see below
**Describe the expected behavior**
generate different numbers 

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
colab: https://colab.research.google.com/drive/1ffXUYw6M7rDokL2EkPqKjna8AD5QEslA?usp=sharing
```python
import tensorflow as tf
import numpy as np

def _h(model,x):
    model(x)

class b(tf.keras.layers.Layer):
    def call(self, inputs):
        if tf.constant(True):
            tf.print(tf.random.uniform([5,]),summarize=-1)
        else:
            pass
        return inputs

tf.random.set_seed(123)
inputs = tf.keras.Input(shape=(784,))

x=b()(inputs)
x=b()(x)
x=b()(x)
outputs=b()(x)

model = tf.keras.Model(inputs=inputs, outputs=outputs)

h=tf.function(_h)

print('first run:')
h(model,tf.constant(np.random.rand(64,784)))
print('second run:')
h(model,tf.constant(np.random.rand(64,784)))
```
output:
first run:
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
second run:
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]

The phenomenon generalize to other related api like `tf.random.categorical` , etc.
I read this issue, https://github.com/tensorflow/tensorflow/issues/33297, and realized that tf random could be buggy, but there is no any warning to new users in the api page.(e.g. https://www.tensorflow.org/api_docs/python/tf/random/uniform) The warning is shown in https://www.tensorflow.org/guide/random_numbers, which is too difficult to notice. The random number docs is inadequate as well, how do I use more complicated api like `tf.random.categorical` from `tf.random.Generator`?

**Comment by Saduf2019** [comment]
@laplacericky 
The issue is using tf.constant in your code, i have changed that and able to produce different sequence with  tf random, please find the [gist here.](https://colab.research.google.com/gist/Saduf2019/1f43ab5329ebae2b53832c829443a02d/untitled606.ipynb).

Kindly open a [stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow) issue for this as it is not a bug or feature request, Please post this kind of support questions at Stackoverflow. There is a big community to support and learn from your questions.
Thanks!

**Comment by laplacericky** [comment]
> @laplacericky
> The issue is using tf.constant in your code, i have changed that and able to produce different sequence with tf random, please find the [gist here.](https://colab.research.google.com/gist/Saduf2019/1f43ab5329ebae2b53832c829443a02d/untitled606.ipynb).
> 
> Kindly open a [stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow) issue for this as it is not a bug or feature request, Please post this kind of support questions at Stackoverflow. There is a big community to support and learn from your questions.
> Thanks!

That is a very common use case that I want to have some comparisons which involves tensor (i.e. tf.cond) inside call function of a custom keras layers. In here in my bug reproduction code, I used `tf.constant(True)` just for the sake of having the if-else clause which will be converted by autograph and I want to have minimal code to reproduce the bug. In reality case, I could be comparing a tensor with another tensor. 

The bug can only be reproduced inside autograph-converted tf.cond which further demonstrates it is a dangerous bug because it sometime works and sometimes don't work. Unless there is clear docs about the situation, more and more people will fall into this trap and thought it is generating random numbers but indeed it generates the same sequence of numbers.

**Comment by Saduf2019** [comment]
@laplacericky 
This is not a bug, please go through the documentation for random seed and open issue at Stackoverflow/tf discussion forum

**Comment by laplacericky** [comment]
@Saduf2019 
Thank you for your reply. Please provide explanations that why this behavior is not considered as bug. I will also provide codes for  several ways to reproduce this behavior and how to avoid this behavior.

**Comment by laplacericky** [comment]
Way 1 to reproduce this behavior
```python
import tensorflow as tf
import numpy as np


def _h(model,x):
    model(x)

class buggy_random(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        self.boolean_var=tf.Variable(True)
        super().__init__(**kwargs)
    def call(self, inputs):
        if self.boolean_var:
          tf.print(tf.random.uniform([5,]))

        return inputs

tf.random.set_seed(123)
inputs = tf.keras.Input(shape=(784,))

x=buggy_random()(inputs)
x=buggy_random()(x)
x=buggy_random()(x)
outputs=buggy_random()(x)

model = tf.keras.Model(inputs=inputs, outputs=outputs)
model.layers[1].boolean_var.assign(False)
model.layers[3].boolean_var.assign(False)


h=tf.function(_h)
print('first run:')
h(model,tf.constant(np.random.rand(64,784)))
print('second run:')
h(model,tf.constant(np.random.rand(64,784)))
'''
first run:
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
second run:
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
'''
```

**Comment by laplacericky** [comment]
Way 2 to reproduce this behavior:
```python
import tensorflow as tf
import numpy as np

def _h(model,x):
    model(x)

class buggy_random(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
    def call(self, inputs):
        for i in tf.range(tf.constant(1)):
          tf.print(tf.random.uniform([5,]))

        return inputs

tf.random.set_seed(123)
inputs = tf.keras.Input(shape=(784,))

x=buggy_random()(inputs)
x=buggy_random()(x)
x=buggy_random()(x)
outputs=buggy_random()(x)

model = tf.keras.Model(inputs=inputs, outputs=outputs)

h=tf.function(_h)
print('first run:')
h(model,tf.constant(np.random.rand(64,784)))
print('second run:')
h(model,tf.constant(np.random.rand(64,784)))

'''
first run:
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
[0.277247906 0.994074821 0.379808426 0.71479249 0.50061965]
second run:
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
[0.176383495 0.109812617 0.334476113 0.66576612 0.116794825]
'''
```

**Comment by laplacericky** [comment]
The new API is the only working randomness API in tensorflow.
```python
import tensorflow as tf
import numpy as np

def _h(model,x):
    model(x)

class fixed_random(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        self.g = tf.random.Generator.from_seed(np.random.randint(2147483647))
        super().__init__(**kwargs)
    def call(self, inputs):
        for i in tf.range(tf.constant(1)):
            tf.print(self.g.uniform([5,]),summarize=-1)
        return inputs

tf.random.set_seed(123)
np.random.seed(88883)
inputs = tf.keras.Input(shape=(784,))

x=fixed_random()(inputs)
x=fixed_random()(x)
x=fixed_random()(x)
outputs=fixed_random()(x)

model = tf.keras.Model(inputs=inputs, outputs=outputs)

h=tf.function(_h)

print('first run')
h(model,tf.constant(np.random.rand(64,784)))
print('second run')
h(model,tf.constant(np.random.rand(64,784)))
'''
first run
[0.898076177 0.814805269 0.685280442 0.624640703 0.637957335]
[0.438643932 0.923692703 0.202350736 0.259410977 0.551186323]
[0.137919307 0.688777804 0.0962672234 0.532677054 0.639819145]
[0.533203602 0.589777708 0.596117854 0.147099137 0.760685086]
second run
[0.391331553 0.65578413 0.0532933474 0.92483747 0.198105812]
[0.0157146454 0.148599625 0.300893545 0.211638093 0.460565567]
[0.310432553 0.377720356 0.230332017 0.0171860456 0.905866146]
[0.927248478 0.930060267 0.906018138 0.391339064 0.362027764]
'''
```

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
